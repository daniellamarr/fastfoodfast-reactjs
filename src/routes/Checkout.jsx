import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { getCartItems, deleteItem } from '../actions/cartActions';
import { placeOrder } from '../actions/orderActions';

/**
 * Checkout
 */
export class Checkout extends Component {
  /**
   * Constructor
   */
  constructor() {
    super();

    this.state = {
      sumOfPrices: [],
    };
  }

  /**
   * Total Price
   * @return {number} Total Price of items in cart
   */
  totalPrice() {
    let totalPrice = 0;
    if (this.props.cart.cart.length > 0) {
      totalPrice = this.state.sumOfPrices.reduce(
        (price, quantity) => price + quantity,
      );
    }

    return totalPrice;
  }

  /**
   * @returns {JSX} jsx
   */
  render() {
    return (
      <div>
        <section id="order">
          <div className="title">
            <h1>Checkout</h1>
            <p>Place your order</p>
          </div>
        </section>
        <section id="checkout">
          <div className="table">
            <table>
              <thead>
                <tr>
                  <th>Product Image</th>
                  <th>Product</th>
                  <th>Quantity</th>
                  <th>Price(N)</th>
                  <th>Total</th>
                  <th>Action</th>
                </tr>
              </thead>
              <tbody id="cartorders">
                {
                  this.props.cart.cart.map((item, index) => {
                    const quantityPrice = item.price * item.quantity;
                    this.state.sumOfPrices.push(quantityPrice);
                    return (
                      <tr key={item.id}>
                        <td>
                          <img
                            src={item.image}
                            alt="Product Image"
                            width="80px"
                          />
                        </td>
                        <td>
                          {item.title}
                        </td>
                        <td>
                          {item.quantity}
                        </td>
                        <td>
                          {item.price}
                        </td>
                        <td>
                          {quantityPrice}
                        </td>
                        <td>
                        <a
                          className="red-text"
                          onClick={() => this.props.deleteItem(index)}
                        >
                          <i className="ti-close"></i>
                        </a>
                        </td>
                      </tr>
                    );
                  })
                }
                <tr className="tabletotal">
                  <td colSpan={4}>
                    Total:
                  </td>
                  <td>
                    {this.totalPrice()}
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
          <br /><br />
          <button
            id="ordernow"
            onClick={() => this.props.placeOrder()}
            disabled={this.props.orders.loadingStatus}
          >
            {this.props.orders.loadingStatus ? 'Loading. . .' : 'Place Order'}
          </button>
        </section>
        <div className="error hide"></div>
      </div>
    );
  }
}

Checkout.propTypes = {
  getCartItems: PropTypes.func.isRequired,
  deleteItem: PropTypes.func.isRequired,
  placeOrder: PropTypes.func.isRequired,
  cart: PropTypes.shape({
    cart: PropTypes.arrayOf(PropTypes.shape({
      title: PropTypes.string.isRequired,
    })),
    noOfItems: PropTypes.number,
    cartStatus: PropTypes.bool,
  }),
  orders: PropTypes.shape({
    loadingStatus: PropTypes.bool,
  }),
};

export const mapStateToProps = state => ({
  cart: state.cart,
  orders: state.orders,
});

export default connect(mapStateToProps,
  { getCartItems, deleteItem, placeOrder })(Checkout);
