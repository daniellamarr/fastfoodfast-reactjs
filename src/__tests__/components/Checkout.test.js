/* eslint-disable no-undef */
import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import { Checkout } from '../../routes/Checkout.jsx';

const middlewares = [thunk];

const mockStore = configureMockStore(middlewares);
const store = mockStore();

let wrapper;

const props = {
  getCartItems: () => {},
  deleteItem: () => {},
  placeOrder: () => {},
  cart: {
    cart: [
      {
        title: 'Hello Test',
        price: 100,
        quantity: 20,
      },
    ],
    noOfItems: 1,
  },
  orders: {
    loadingStatus: false,
  },
};

beforeEach(() => {
  wrapper = mount(
    <Provider store={store}>
      <Checkout {...props} />
    </Provider>,
  );
});
afterEach(() => {
  wrapper.unmount();
});

describe('Profile Component', () => {
  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('click delete button', () => {
    wrapper.find('.red-text').simulate('click');
  });

  it('click place order button', () => {
    wrapper.find('#ordernow').simulate('click');
  });

  it('should simulate an empty cart', () => {
    const props2 = { ...props, cart: { cart: [] } };
    wrapper = mount(<Checkout {...props2} />);
  });
});
