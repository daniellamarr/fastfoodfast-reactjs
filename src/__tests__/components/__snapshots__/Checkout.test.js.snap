// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile Component should match snapshot 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Checkout
    cart={
      Object {
        "cart": Array [
          Object {
            "title": "Hello Test",
          },
        ],
        "noOfItems": 1,
      }
    }
    deleteItem={[Function]}
    getCartItems={[Function]}
    orders={
      Object {
        "loadingStatus": false,
      }
    }
    placeOrder={[Function]}
  >
    <div>
      <section
        id="order"
      >
        <div
          className="title"
        >
          <h1>
            Checkout
          </h1>
          <p>
            Place your order
          </p>
        </div>
      </section>
      <section
        id="checkout"
      >
        <div
          className="table"
        >
          <table>
            <thead>
              <tr>
                <th>
                  Product Image
                </th>
                <th>
                  Product
                </th>
                <th>
                  Quantity
                </th>
                <th>
                  Price(N)
                </th>
                <th>
                  Total
                </th>
                <th>
                  Action
                </th>
              </tr>
            </thead>
            <tbody
              id="cartorders"
            >
              <tr
                key="0"
              >
                <td>
                  <img
                    alt="Product Image"
                    width="80px"
                  />
                </td>
                <td>
                  Hello Test
                </td>
                <td />
                <td />
                <td>
                  NaN
                </td>
                <td>
                  <a
                    className="red-text"
                    onClick={[Function]}
                  >
                    <i
                      className="ti-close"
                    />
                  </a>
                </td>
              </tr>
              <tr
                className="tabletotal"
              >
                <td
                  colSpan={4}
                >
                  Total:
                </td>
                <td>
                  NaN
                </td>
              </tr>
            </tbody>
          </table>
        </div>
        <br />
        <br />
        <button
          disabled={false}
          id="ordernow"
          onClick={[Function]}
        >
          Place Order
        </button>
      </section>
      <div
        className="error hide"
      />
    </div>
  </Checkout>
</Provider>
`;
